File: internal/routes/app_updates.go
package routes

import (
	"bytes"
	"context"
	"crypto/rand"
	"crypto/sha256"
	"encoding/binary"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"time"

	apptypes "github.com/ClashKingInc/ClashKingAPI/internal/utils"
	"github.com/gofiber/fiber/v2"
	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
)

const installationHeader = "x-clashking-installation"

var (
	appUpdateVersionPattern     = regexp.MustCompile(`^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(-beta)?$`)
	appUpdateSignaturePattern   = regexp.MustCompile(`^sig="[A-Za-z0-9+/]+={0,2}", keyid="main"$`)
	appUpdateRollbackKeyPattern = regexp.MustCompile(`^rollbacks\/(beta|production)\/[^/]+\/[^/]+\/(ios|android)-[0-9a-f-]{36}\.json$`)
)

type appUpdateDB interface {
	QueryRow(context.Context, string, ...any) pgx.Row
}

type appUpdateMarkerLoader interface {
	Load(context.Context, string, string) (appUpdateRelease, error)
	LoadRollback(context.Context, string) (appUpdatePlatform, error)
}

type appUpdateHTTPMarkerLoader struct {
	origin string
	client *http.Client
}

type appUpdateRelease struct {
	SchemaVersion   int                                `json:"schemaVersion"`
	Version         string                             `json:"version"`
	Track           string                             `json:"track"`
	Type            string                             `json:"type"`
	Platforms       map[string]appUpdatePlatform       `json:"platforms"`
	RollbackTargets map[string]appUpdateRollbackTarget `json:"rollbackTargets"`
}

type appUpdateRollbackTarget struct {
	Type      string                               `json:"type"`
	Platforms map[string]appUpdateRollbackPlatform `json:"platforms"`
}

type appUpdateRollbackPlatform struct {
	RuntimeVersion string `json:"runtimeVersion"`
	Key            string `json:"key"`
}

type appUpdatePlatform struct {
	RuntimeVersion string          `json:"runtimeVersion"`
	Manifest       json.RawMessage `json:"manifest"`
	Signature      string          `json:"signature"`
}

type appUpdateChannelState struct {
	ActiveVersion         string
	RollbackTargetVersion string
	BasisPoints           int
	Paused                bool
	From                  *int
	To                    *int
	StartsAt              *time.Time
	EndsAt                *time.Time
}

func appUpdateManifest(a apptypes.Deps) fiber.Handler {
	var database appUpdateDB
	if a.Store != nil {
		database = a.Store.SQL
	}
	loader := appUpdateHTTPMarkerLoader{
		origin: a.Config.AppUpdatesR2Origin,
		client: &http.Client{Timeout: 10 * time.Second},
	}
	return appUpdateManifestHandler(database, loader, time.Now)
}

func appUpdateManifestHandler(database appUpdateDB, loader appUpdateMarkerLoader, now func() time.Time) fiber.Handler {
	return func(c *fiber.Ctx) error {
		if database == nil {
			return apptypes.Error(fiber.StatusServiceUnavailable, "Update service is not configured")
		}
		if c.Get("expo-protocol-version") != "1" {
			return apptypes.Error(fiber.StatusBadRequest, "Unsupported Expo Updates protocol version")
		}

		platform := strings.ToLower(strings.TrimSpace(c.Get("expo-platform")))
		if platform != "ios" && platform != "android" {
			return apptypes.Error(fiber.StatusBadRequest, "Unsupported update platform")
		}
		runtimeVersion := strings.TrimSpace(c.Get("expo-runtime-version"))
		if runtimeVersion == "" || len(runtimeVersion) > 200 {
			return apptypes.Error(fiber.StatusBadRequest, "Invalid Expo runtime version")
		}
		channel := strings.ToLower(strings.TrimSpace(c.Get("expo-channel-name")))
		if channel == "" {
			channel = "production"
		}
		if channel != "beta" && channel != "production" {
			return apptypes.Error(fiber.StatusBadRequest, "Unsupported update channel")
		}

		token, err := installationToken(c.Get(installationHeader))
		if err != nil {
			return apptypes.Error(fiber.StatusBadRequest, "Invalid installation cohort token")
		}
		c.Set("expo-server-defined-headers", installationHeader+"=\""+token+"\"")
		c.Set("expo-protocol-version", "1")
		c.Set("expo-sfv-version", "0")
		c.Set(fiber.HeaderCacheControl, "private, max-age=0")

		state, err := loadAppUpdateChannel(c.UserContext(), database, channel, platform, runtimeVersion)
		if errors.Is(err, pgx.ErrNoRows) {
			return c.SendStatus(fiber.StatusNoContent)
		}
		if err != nil {
			return apptypes.Error(fiber.StatusServiceUnavailable, "Update service is unavailable")
		}
		currentUpdateID := nullableUpdateID(c.Get("expo-current-update-id"))
		if state.ActiveVersion == "" || state.Paused {
			return c.SendStatus(fiber.StatusNoContent)
		}

		basisPoints := effectiveRolloutBasisPoints(state, now().UTC())
		if rolloutBucket(state.ActiveVersion, token) >= basisPoints {
			return c.SendStatus(fiber.StatusNoContent)
		}

		release, err := loader.Load(c.UserContext(), channel, state.ActiveVersion)
		if err != nil {
			return apptypes.Error(fiber.StatusServiceUnavailable, "Update release is unavailable")
		}
		if release.SchemaVersion != 1 || release.Type != "ota" || release.Track != channel || release.Version != state.ActiveVersion || !validAppUpdateReleaseVersion(release.Version, channel, release.Type) {
			return apptypes.Error(fiber.StatusServiceUnavailable, "Update release marker is invalid")
		}
		update, servedVersion, err := selectAppUpdatePlatform(c.UserContext(), loader, release, state, platform)
		if err != nil || update.RuntimeVersion != runtimeVersion || len(update.Manifest) == 0 || !validAppUpdateSignature(update.Signature) {
			return apptypes.Error(fiber.StatusServiceUnavailable, "Update release marker is invalid")
		}
		update.Manifest, err = compactAppUpdateManifest(update.Manifest)
		if err != nil {
			return apptypes.Error(fiber.StatusServiceUnavailable, "Update manifest is invalid")
		}

		var manifestID struct {
			ID             string `json:"id"`
			RuntimeVersion string `json:"runtimeVersion"`
			Metadata       struct {
				Version      string `json:"version"`
				RollbackFrom string `json:"rollbackFrom"`
			} `json:"metadata"`
		}
		if json.Unmarshal(update.Manifest, &manifestID) != nil || manifestID.ID == "" || manifestID.RuntimeVersion != runtimeVersion || manifestID.Metadata.Version != servedVersion {
			return apptypes.Error(fiber.StatusServiceUnavailable, "Update manifest is invalid")
		}
		if state.RollbackTargetVersion != "" && manifestID.Metadata.RollbackFrom != state.ActiveVersion {
			return apptypes.Error(fiber.StatusServiceUnavailable, "Update manifest is invalid")
		}
		if currentUpdateID != nil && manifestID.ID == currentUpdateID.String() {
			return c.SendStatus(fiber.StatusNoContent)
		}
		return sendAppUpdateManifest(c, update.Manifest, update.Signature)
	}
}

func selectAppUpdatePlatform(ctx context.Context, loader appUpdateMarkerLoader, release appUpdateRelease, state appUpdateChannelState, platform string) (appUpdatePlatform, string, error) {
	if state.RollbackTargetVersion == "" {
		update, ok := release.Platforms[platform]
		if !ok {
			return appUpdatePlatform{}, "", errors.New("platform is not present in release")
		}
		return update, release.Version, nil
	}
	if state.RollbackTargetVersion == state.ActiveVersion {
		return appUpdatePlatform{}, "", errors.New("rollback target matches active release")
	}
	target, ok := release.RollbackTargets[state.RollbackTargetVersion]
	if !ok || !validAppUpdateReleaseVersion(state.RollbackTargetVersion, release.Track, target.Type) {
		return appUpdatePlatform{}, "", errors.New("rollback target is invalid")
	}
	reference, ok := target.Platforms[platform]
	expectedPrefix := fmt.Sprintf("rollbacks/%s/%s/%s/%s-", release.Track, release.Version, state.RollbackTargetVersion, platform)
	if !ok || !strings.HasPrefix(reference.Key, expectedPrefix) || !appUpdateRollbackKeyPattern.MatchString(reference.Key) {
		return appUpdatePlatform{}, "", errors.New("rollback platform is invalid")
	}
	update, err := loader.LoadRollback(ctx, reference.Key)
	if err != nil || update.RuntimeVersion != reference.RuntimeVersion {
		return appUpdatePlatform{}, "", errors.New("rollback descriptor is invalid")
	}
	return update, state.RollbackTargetVersion, nil
}

func validAppUpdateSignature(value string) bool {
	return appUpdateSignaturePattern.MatchString(value)
}

func compactAppUpdateManifest(manifest json.RawMessage) (json.RawMessage, error) {
	var compact bytes.Buffer
	if err := json.Compact(&compact, manifest); err != nil {
		return nil, err
	}
	return json.RawMessage(compact.String()), nil
}

func validAppUpdateReleaseVersion(version, channel, releaseType string) bool {
	match := appUpdateVersionPattern.FindStringSubmatch(version)
	if match == nil || (channel == "beta") != strings.HasSuffix(version, "-beta") {
		return false
	}
	return (releaseType == "native" && match[3] == "0") || (releaseType == "ota" && match[3] != "0")
}

func loadAppUpdateChannel(ctx context.Context, database appUpdateDB, channel, platform, runtimeVersion string) (appUpdateChannelState, error) {
	var state appUpdateChannelState
	err := database.QueryRow(ctx, `SELECT COALESCE(active_version, ''), COALESCE(rollback_target_version, ''), rollout_basis_points, paused,
        rollout_from_basis_points, rollout_to_basis_points, rollout_starts_at, rollout_ends_at
      FROM app_update_channels WHERE channel = $1 AND platform = $2 AND runtime_version = $3`,
		channel, platform, runtimeVersion,
	).Scan(&state.ActiveVersion, &state.RollbackTargetVersion, &state.BasisPoints, &state.Paused, &state.From, &state.To, &state.StartsAt, &state.EndsAt)
	return state, err
}

func installationToken(value string) (string, error) {
	value = strings.TrimSpace(value)
	if value != "" {
		decoded, err := hex.DecodeString(value)
		if err != nil || len(decoded) != 16 {
			return "", errors.New("invalid token")
		}
		return strings.ToLower(value), nil
	}
	b := make([]byte, 16)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return hex.EncodeToString(b), nil
}

func nullableUpdateID(value string) *uuid.UUID {
	id, err := uuid.Parse(strings.TrimSpace(value))
	if err != nil {
		return nil
	}
	return &id
}

func effectiveRolloutBasisPoints(state appUpdateChannelState, at time.Time) int {
	if state.From == nil || state.To == nil || state.StartsAt == nil || state.EndsAt == nil {
		return clampBasisPoints(state.BasisPoints)
	}
	if !at.After(*state.StartsAt) {
		return clampBasisPoints(*state.From)
	}
	if !at.Before(*state.EndsAt) {
		return clampBasisPoints(*state.To)
	}
	progress := float64(at.Sub(*state.StartsAt)) / float64(state.EndsAt.Sub(*state.StartsAt))
	value := float64(*state.From) + float64(*state.To-*state.From)*progress
	return clampBasisPoints(int(value + 0.5))
}

func clampBasisPoints(value int) int {
	if value < 0 {
		return 0
	}
	if value > 10000 {
		return 10000
	}
	return value
}

func rolloutBucket(releaseID, token string) int {
	hash := sha256.Sum256([]byte(releaseID + ":" + token))
	return int(binary.BigEndian.Uint64(hash[:8]) % 10000)
}

func (l appUpdateHTTPMarkerLoader) Load(ctx context.Context, channel, version string) (appUpdateRelease, error) {
	if l.origin == "" {
		return appUpdateRelease{}, errors.New("R2 update origin is not configured")
	}
	if channel != "beta" && channel != "production" {
		return appUpdateRelease{}, errors.New("invalid channel")
	}
	if !appUpdateVersionPattern.MatchString(version) {
		return appUpdateRelease{}, errors.New("invalid version")
	}
	markerURL := strings.TrimRight(l.origin, "/") + "/releases/" + url.PathEscape(channel) + "/" + url.PathEscape(version) + "/release.json"
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, markerURL, nil)
	if err != nil {
		return appUpdateRelease{}, err
	}
	res, err := l.client.Do(req)
	if err != nil {
		return appUpdateRelease{}, err
	}
	defer res.Body.Close()
	if res.StatusCode != http.StatusOK {
		return appUpdateRelease{}, fmt.Errorf("marker returned %s", res.Status)
	}
	body, err := io.ReadAll(io.LimitReader(res.Body, 2<<20))
	if err != nil {
		return appUpdateRelease{}, err
	}
	var release appUpdateRelease
	if err := json.Unmarshal(body, &release); err != nil {
		return appUpdateRelease{}, err
	}
	return release, nil
}

func (l appUpdateHTTPMarkerLoader) LoadRollback(ctx context.Context, key string) (appUpdatePlatform, error) {
	if l.origin == "" {
		return appUpdatePlatform{}, errors.New("R2 update origin is not configured")
	}
	if !appUpdateRollbackKeyPattern.MatchString(key) {
		return appUpdatePlatform{}, errors.New("invalid rollback descriptor key")
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, strings.TrimRight(l.origin, "/")+"/"+key, nil)
	if err != nil {
		return appUpdatePlatform{}, err
	}
	res, err := l.client.Do(req)
	if err != nil {
		return appUpdatePlatform{}, err
	}
	defer res.Body.Close()
	if res.StatusCode != http.StatusOK {
		return appUpdatePlatform{}, fmt.Errorf("rollback descriptor returned %s", res.Status)
	}
	body, err := io.ReadAll(io.LimitReader(res.Body, 2<<20))
	if err != nil {
		return appUpdatePlatform{}, err
	}
	var update appUpdatePlatform
	if err := json.Unmarshal(body, &update); err != nil {
		return appUpdatePlatform{}, err
	}
	return update, nil
}

func sendAppUpdateManifest(c *fiber.Ctx, manifest json.RawMessage, signature string) error {
	boundaryBytes := make([]byte, 12)
	if _, err := rand.Read(boundaryBytes); err != nil {
		return err
	}
	boundary := "expo-" + hex.EncodeToString(boundaryBytes)
	var body strings.Builder
	body.WriteString("--" + boundary + "\r\n")
	body.WriteString("content-disposition: form-data; name=\"manifest\"\r\n")
	body.WriteString("content-type: application/json; charset=utf-8\r\n")
	if signature != "" {
		body.WriteString("expo-signature: " + signature + "\r\n")
	}
	body.WriteString("\r\n")
	body.Write(manifest)
	body.WriteString("\r\n--" + boundary + "--\r\n")
	c.Set(fiber.HeaderContentType, "multipart/mixed; boundary="+boundary)
	return c.SendString(body.String())
}

File: workers/api/src/app-updates.ts
import { Effect } from "effect"
import { SqlClient } from "effect/unstable/sql"

import { DatabaseFailure, InvalidRequest, UpstreamUnavailable } from "./errors.js"
import type { WorkerBindings } from "./environment.js"

interface ChannelRow {
  readonly active_version: string
  readonly paused: boolean
  readonly rollback_target_version: string
  readonly rollout_basis_points: number
  readonly rollout_ends_at: Date | string | null
  readonly rollout_from_basis_points: number | null
  readonly rollout_starts_at: Date | string | null
  readonly rollout_to_basis_points: number | null
}

interface UpdatePlatform {
  readonly manifest: unknown
  readonly runtimeVersion: string
  readonly signature: string
}

interface RollbackTarget {
  readonly platforms: Partial<Record<"android" | "ios", { readonly key: string; readonly runtimeVersion: string }>>
  readonly type: string
}

interface UpdateRelease {
  readonly platforms: Partial<Record<"android" | "ios", UpdatePlatform>>
  readonly rollbackTargets?: Readonly<Record<string, RollbackTarget>>
  readonly schemaVersion: number
  readonly track: string
  readonly type: string
  readonly version: string
}

const versionPattern = /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(-beta)?$/u
const signaturePattern = /^sig="[A-Za-z0-9+/]+={0,2}", keyid="main"$/u
const rollbackKeyPattern = /^rollbacks\/(beta|production)\/[^/]+\/[^/]+\/(ios|android)-[0-9a-f-]{36}\.json$/u

const invalid = (message: string) => new InvalidRequest({ message })

const installationToken = (value: string | null): Effect.Effect<string, InvalidRequest> => {
  const token = value?.trim().toLowerCase() ?? ""
  if (token.length > 0) {
    return /^[0-9a-f]{32}$/u.test(token)
      ? Effect.succeed(token)
      : Effect.fail(invalid("Invalid installation cohort token"))
  }
  const bytes = crypto.getRandomValues(new Uint8Array(16))
  return Effect.succeed([...bytes].map((byte) => byte.toString(16).padStart(2, "0")).join(""))
}

const validVersion = (version: string, channel: string, releaseType: string): boolean => {
  const match = versionPattern.exec(version)
  if (match === null || (channel === "beta") !== version.endsWith("-beta")) return false
  return releaseType === "native" ? match[3] === "0" : releaseType === "ota" && match[3] !== "0"
}

const basisPoints = (state: ChannelRow, now: Date): number => {
  const clamp = (value: number) => Math.max(0, Math.min(10_000, value))
  if (state.rollout_from_basis_points === null || state.rollout_to_basis_points === null ||
      state.rollout_starts_at === null || state.rollout_ends_at === null) {
    return clamp(state.rollout_basis_points)
  }
  const start = new Date(state.rollout_starts_at).valueOf()
  const end = new Date(state.rollout_ends_at).valueOf()
  if (now.valueOf() <= start) return clamp(state.rollout_from_basis_points)
  if (now.valueOf() >= end) return clamp(state.rollout_to_basis_points)
  const progress = (now.valueOf() - start) / (end - start)
  return clamp(Math.round(state.rollout_from_basis_points +
    (state.rollout_to_basis_points - state.rollout_from_basis_points) * progress))
}

const rolloutBucket = (version: string, token: string) => Effect.promise(async () => {
  const digest = await crypto.subtle.digest("SHA-256", new TextEncoder().encode(`${version}:${token}`))
  const view = new DataView(digest)
  return Number(view.getBigUint64(0) % 10_000n)
})

const loadJson = <A>(bucket: R2Bucket, key: string): Effect.Effect<A, UpstreamUnavailable> =>
  Effect.tryPromise({
    try: async () => {
      const object = await bucket.get(key)
      if (object === null || object.size > 2 * 1024 * 1024) throw new Error(`Missing or oversized R2 object ${key}`)
      return await object.json() as A
    },
    catch: (cause) => new UpstreamUnavailable({ cause, message: "Update release is unavailable" }),
  })

const noUpdate = (token: string): Response => new Response(null, {
  status: 204,
  headers: updateHeaders(token),
})

const updateHeaders = (token: string): Headers => new Headers({
  "cache-control": "private, max-age=0",
  "expo-protocol-version": "1",
  "expo-server-defined-headers": `x-clashking-installation="${token}"`,
  "expo-sfv-version": "0",
})

const manifestResponse = (manifest: unknown, signature: string, token: string): Response => {
  const random = crypto.getRandomValues(new Uint8Array(12))
  const boundary = `expo-${[...random].map((byte) => byte.toString(16).padStart(2, "0")).join("")}`
  const encoded = JSON.stringify(manifest)
  const body = `--${boundary}\r\ncontent-disposition: form-data; name="manifest"\r\ncontent-type: application/json; charset=utf-8\r\nexpo-signature: ${signature}\r\n\r\n${encoded}\r\n--${boundary}--\r\n`
  const headers = updateHeaders(token)
  headers.set("content-type", `multipart/mixed; boundary=${boundary}`)
  return new Response(body, { status: 200, headers })
}

export const serveAppUpdateManifest = (
  request: Request,
  bindings: WorkerBindings,
): Effect.Effect<Response, DatabaseFailure | InvalidRequest | UpstreamUnavailable, SqlClient.SqlClient> => Effect.gen(function* () {
  if (request.headers.get("expo-protocol-version") !== "1") {
    return yield* invalid("Unsupported Expo Updates protocol version")
  }
  const platform = request.headers.get("expo-platform")?.trim().toLowerCase()
  if (platform !== "ios" && platform !== "android") return yield* invalid("Unsupported update platform")
  const runtimeVersion = request.headers.get("expo-runtime-version")?.trim() ?? ""
  if (runtimeVersion.length === 0 || runtimeVersion.length > 200) return yield* invalid("Invalid Expo runtime version")
  const channel = request.headers.get("expo-channel-name")?.trim().toLowerCase() || "production"
  if (channel !== "beta" && channel !== "production") return yield* invalid("Unsupported update channel")
  const token = yield* installationToken(request.headers.get("x-clashking-installation"))
  const sql = yield* SqlClient.SqlClient
  const states = yield* sql<ChannelRow>`
    SELECT COALESCE(active_version, '') AS active_version,
      COALESCE(rollback_target_version, '') AS rollback_target_version,
      rollout_basis_points, paused, rollout_from_basis_points, rollout_to_basis_points,
      rollout_starts_at, rollout_ends_at
    FROM app_update_channels
    WHERE channel = ${channel} AND platform = ${platform} AND runtime_version = ${runtimeVersion}
  `.pipe(Effect.mapError((cause) => new DatabaseFailure({ cause, message: "Update service is unavailable" })))
  const state = states[0]
  if (state === undefined || state.active_version.length === 0 || state.paused) return noUpdate(token)
  if ((yield* rolloutBucket(state.active_version, token)) >= basisPoints(state, new Date())) return noUpdate(token)

  const release = yield* loadJson<UpdateRelease>(
    bindings.APP_UPDATES,
    `releases/${channel}/${state.active_version}/release.json`,
  )
  if (release.schemaVersion !== 1 || release.type !== "ota" || release.track !== channel ||
      release.version !== state.active_version || !validVersion(release.version, channel, release.type)) {
    return yield* new UpstreamUnavailable({ cause: release, message: "Update release marker is invalid" })
  }
  let update = release.platforms[platform]
  let servedVersion = release.version
  if (state.rollback_target_version.length > 0) {
    if (state.rollback_target_version === state.active_version) {
      return yield* new UpstreamUnavailable({ cause: state, message: "Update rollback marker is invalid" })
    }
    const target = release.rollbackTargets?.[state.rollback_target_version]
    const reference = target?.platforms[platform]
    const prefix = `rollbacks/${channel}/${release.version}/${state.rollback_target_version}/${platform}-`
    if (target === undefined || reference === undefined || !validVersion(state.rollback_target_version, channel, target.type) ||
        !reference.key.startsWith(prefix) || !rollbackKeyPattern.test(reference.key)) {
      return yield* new UpstreamUnavailable({ cause: target, message: "Update rollback marker is invalid" })
    }
    update = yield* loadJson<UpdatePlatform>(bindings.APP_UPDATES, reference.key)
    servedVersion = state.rollback_target_version
    if (update.runtimeVersion !== reference.runtimeVersion) {
      return yield* new UpstreamUnavailable({ cause: update, message: "Update rollback marker is invalid" })
    }
  }
  if (update === undefined || update.runtimeVersion !== runtimeVersion || !signaturePattern.test(update.signature) ||
      typeof update.manifest !== "object" || update.manifest === null) {
    return yield* new UpstreamUnavailable({ cause: update, message: "Update release marker is invalid" })
  }
  const manifest = update.manifest as { readonly id?: unknown; readonly runtimeVersion?: unknown; readonly metadata?: unknown }
  const metadata = manifest.metadata as { readonly rollbackFrom?: unknown; readonly version?: unknown } | undefined
  if (typeof manifest.id !== "string" || manifest.runtimeVersion !== runtimeVersion || metadata?.version !== servedVersion ||
      (state.rollback_target_version.length > 0 && metadata.rollbackFrom !== state.active_version)) {
    return yield* new UpstreamUnavailable({ cause: manifest, message: "Update manifest is invalid" })
  }
  if (request.headers.get("expo-current-update-id")?.trim() === manifest.id) return noUpdate(token)
  return manifestResponse(manifest, update.signature, token)
}).pipe(Effect.withSpan("AppUpdates.manifest"))

export const appUpdateInternals = { basisPoints, installationToken, rolloutBucket, validVersion }
